client/src/App.js

```diff
@@ -69,20 +81,37 @@ class App extends Component {
     };
 
     changeLang = (lang) => {
-        const uiContext = { ...this.state.uiContext };
+        const { uiContext } = this.state;
+        const { dataContext } = this.state;
+        const { userName } = dataContext;
+        const { firstName, lastName } = userName;
+
         uiContext.lang = lang;
         uiContext.dictionary = dictionaryList[lang];
-        this.setState({ uiContext });
+
+        dataContext.userName = {
+            ...userName,
```

Comment:
First it was a bit confusing me, partly because of the different syntax in "setUsername".
I would replace "...userName" with "firstName, lastName".
Because: currently it would cause a problem to swap this line & the next line.
And there is duplication behind the scenes.
If replaced, there is no need to declare & initialize userName.
With "firstName, lastName" option we can utilize JS syntax for Object initialization.